### ISSUE TITLE
Drop Scala 2.12 for 1.0?

### ISSUE THREAD

**Issue by armanbilge** [http4s/http4s#5289]
I know this is a potential landmine, so please note the ? and that this is an issue and not a PR 😉

I wanted to spin this elephant out of https://github.com/http4s/http4s/discussions/5245#discussioncomment-1363798. IMHO, why:
* anyone who wants 2.12 support can take their pick from 0.21/0.22/0.23 etc.
* upstream pressure: https://github.com/circe/circe/pull/1818
* ecosystem catch-up: [SPARK-25075](https://issues.apache.org/jira/browse/SPARK-25075)
* singleton/literal types for https://github.com/http4s/http4s/pull/5093
* as noted by @bplommer, much better type inference, helpful for:
  * https://github.com/http4s/http4s/pull/4509
  * https://github.com/http4s/http4s/pull/5229
  * https://github.com/http4s/http4s/pull/5251

**Comment by bplommer** [comment]
How about this:

1. Drop 2.12 from `main`
2. Develop against 2.13 until release candidate n.
3. Try to backport RCn to 2.12 while maintaining forwards binary (not source) compatibility.
  a. If this proves feasible, release the backported version as 1.0.0 and RCn as 1.1.0 - provide bug fixes only for 1.0.x
  b. If this doesn't prove feasible, 2.12 was holding http4s back and needs to be dropped.

**Comment by rossabaker** [comment]
I mostly like Ben's idea.  It's an organic way to make the decision.  I don't love supporting two 1.x lines out of the gate along with an 0.23 we'll have to keep viable while people migrate.  I also don't think the backport will be feasible if we stop dragging 2.12 along now.

**Comment by armanbilge** [comment]
I agree, it would be very organic but I'm also doubtful about the feasibility of a backport. Doesn't simply merging your PR using singleton types already put 2.12 support in a very awkward place?

I don't think it's _impossible_, but would probably require splitting into 2.12/2.13-3 sources. So then we're talking about two branches, with two separate sets of sources in the 1.0 branch (and all the ugliness involved with split files).

**Comment by bplommer** [comment]
> I agree, it would be very organic but I'm also doubtful about the feasibility of a backport. Doesn't simply merging your PR using singleton types already put 2.12 support in a very awkward place?

That PR is against 0.22 and doesn't break bincompat, so yes, it puts 2.12 in an awkward place but it doesn't rule out a forward-binary-compatible backport.

I do accept the broader point that following my proposal is *very likely*, albeit not certain, to result in 1.0 lacking any 2.12 support.

**Comment by armanbilge** [comment]
> I do accept the broader point that following my proposal is very likely, albeit not certain, to result in 1.0 lacking any 2.12 support.

Sure, I mean if using the party line "We're developing against 2.13 but we'll backport to 2.12" helps us to sleep better at night for now, I'm all for it. FWIW I think CE3 was developed like this so it's a believable strategy 😉

**Comment by ChristopherDavenport** [comment]
I would like a plan for sbt other than operating on an abandoned version. Which is a reason why many of us still need 2.12 support independent of spark. 

I still support plugins that use http4s in them for 2.12.  For example 

https://github.com/davenverse/sbt-github-download

**Comment by armanbilge** [comment]
That's right, sbt is on 2.12. 😕 I don't know what the plans are for that.

> operating on an abandoned version

I guess this is a race between when 0.23 sunsets and sbt updates to a new Scala?

**Comment by keynmol** [comment]
I feel like this question largely depends on the timelines for 1.0.

The general feel in the ecosystem has been "support two last major versions of the language". Problem is, Scala 3 in its current shape is not complete enough to be a major version in the area where http4s is used most widely - stable production systems. Tooling, lack of all the `-Xwarn` parameters, etc., make Scala 3 not ready yet to shift the window on the `2.12, 2.13, 3.x` line. It's getting there, with very good speed, but it's not there yet.

Pulling the trigger on 1.x is probably still ways off (?correct me if I'm operating under the wrong assumption) because it's not a decision to be taken lightly - Typelevel ecosystem has been **exceptional** in its bincompat guarantees, and releasing 1.x lightly to bump it up to 2.x a month later is probably not what people are hoping to do. I've said it on discord before - whatever time http4s maintainers/contributors need to get it **right** (a very subjective, maintainer focused word), is a time well spent.

If the 1.x is still a year away, then dropping 2.12 out of the gate is reasonable. By that time Scala 3 will truly become a fully fledged, fully supported major version.

If 1.x is releasing in less than 6 months time, then dropping 2.12 is likely alienating people ready to switch to stable APIs of http4s 1.x, people hoping to reduce the 4-branch churn.

_TL;DR in a true sense of the words, there are only 2 major adopted Scala versions right now - 2.12 and 2.13, so http4s 1.x should support both, unless it releases a year from now, when that situation is likely to change_

**Comment by jenshalm** [comment]
@armanbilge, @ChristopherDavenport : Ideally there would be some feedback and alerting for the sbt maintainers before dropping 2.12 support in major libraries as it's a bit of a catch-22 situation. I've seen an older discussion, probably early this year, where sbt maintainers claimed that they want to remain binary compatible as long as possible, until either the core team stops maintaining 2.12 or the community starts dropping it. If the latter would start to happen now, it's better they know about it and maybe reconsider their own timeline, too.

**Comment by armanbilge** [comment]
> whatever time http4s maintainers/contributors need to get it right (a very subjective, maintainer focused word), is a time well spent.

IMHO, getting it "right" for 1.0 should mean unshackling ourselves from 2.12, assuming we agree there are very real benefits that are gained from the non-2.12 features.

> If the latter would start to happen now

Assuming that 1.0 is at least one year away, and that 0.23 won't be sunset for _at least_ one year after 1.0 drops, that gives SBT two years to move past 2.12.

**Comment by rossabaker** [comment]
I don't have a timeline for 1.0, but I'm sweating profusely just keeping up with PRs.  Big picture items that have been stalled for months include server responses in resource and the entity refactor.  I have only started going through the model vs. the HTTP Semantics spec to ensure that we're positioned to implement HTTP/2 and HTTP/3.  Our URI is less of a toxic dump than it was in 0.22, but still is The Shame of Http4s.  I don't believe 1.0 is imminent, and if anyone wants to speed it up, I'll just casually mention that reviews are a bottleneck.

Also, once a dependency becomes established in the SBT ecosystem, I'd be hesitant to upgrade it.  Once someone else writes a popular plugin on the same library, you're tied at the hip or can't both participate in the same build.  I would recommend keeping those on 0.23 through the sbt-1.x era even if http4s-1.0 were tagged this afternoon.  Classpath conservatism is essential in the SBT space.

**Comment by bplommer** [comment]
> If 1.x is releasing in less than 6 months time, then dropping 2.12 is likely alienating people ready to switch to stable APIs of http4s 1.x, people hoping to reduce the 4-branch churn.

I'm just throwing ideas around here, but would it make a difference if the 0.23 series were rebadged as 1.0 to signal commitment to ongoing support for it, and `main` became 2.x? I can understand people who are stuck on 2.12 not wanting to feel abandoned, but I also don't think it's reasonable to hold everyone else back to avoid them feeling like they're missing out.

**Comment by rossabaker** [comment]
We've had so many confusing rebrandings on 1.0, I'm reluctant to do another.  With hindsight on contributor availability last spring and Scala 3 adoption, I think we'd have arranged things differently, but people are getting used to the present version matrix.

**Comment by armanbilge** [comment]
> 0.23 series were rebadged as 1.0 to signal commitment to ongoing support for it

What if we put an LTS badge next to it or something?

**Comment by rossabaker** [comment]
We could, but we've never defined what that means for us.  Everything has historically been "some months, about a year, after the next stable release."

**Comment by armanbilge** [comment]
I think if 1.0 drops 2.12, then the definition of "LTS" would 0.23 would be something like "as long as 2.12 is still kicking".

**Comment by armanbilge** [comment]
Done in https://github.com/http4s/http4s/pull/5337.
